pkg/apis/audit/types.go

```diff
@@ -0,0 +1,166 @@
+	// +optional
+	ObjectRef *api.ObjectReference
+	// The response status, populated even when the ResponseObject is not a Status type.
+	// For successful responses, this will only include the Code and StatusSuccess.
+	// For non-status type error responses, this will be auto-populated with the error Message.
+	// +optional
+	ResponseStatus *metav1.Status
+
+	// RequestObject logged before admission. Omitted for non-resource requests.
+	// Only logged at RequestObject Level and higher.
+	// +optional
+	RequestObject runtime.Object
+	// ResponseObject logged before admission. Omitted for non-resource requests.
+	// Only logged at ResponseObject Level and higher.
+	// Omitted if the response is a metav1.Status object (see ResponseStatus above).
+	// +optional
+	ResponseObject runtime.Object
+}
+
+// EventList is a list of audit Events.
+type EventList struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ListMeta
+
+	Events []Event
+}
+
+// Policy defines the configuration of audit logging, and the rules for how different request
+// categories are logged.
+type Policy struct {
+	metav1.TypeMeta
+	// +optional
+	metav1.ListMeta
+
+	// Rules specify the audit Level a request should be recorded at.
+	// A request may match multiple rules, in which case the FIRST matching rule is used.
+	// The default audit level is None, but can be overridden by a catch-all rule at the end of the list.
+	Rules []PolicyRule
+}
+
+// PolicyRule maps requests based off metadata to an audit Level.
+// Requests must match the rules of every field (an intersection of rules).
+type PolicyRule struct {
+	// The Level that requests matching this rule are recorded at.
+	Level Level
+
+	// The users (by authenticated user name) this rule applies to.
+	// An empty list implies every user.
+	// +optional
+	Users []string
+	// The user groups this rule applies to. If a user is considered matching
+	// if the are a member of any of these groups
+	// An empty list implies every user group.
+	// +optional
+	UserGroups []string
+
+	// The verbs that match this rule.
+	// An empty list implies every verb.
+	// +optional
+	Verbs []string
+
+	// Rules can apply to API resources (such as "pods" or "secrets"),
+	// non-resource URL paths (such as "/api"), or neither, but not both.
+	// If neither is specified, the rule is treated as a default for all URLs.
+
+	// APIGroups is the name of the APIGroup that contains the resources ("" for core).
+	// If multiple API groups are specified, any action requested against one of the
+	// enumerated resources in any of the listed API groups will be allowed.
+	// Any empty list implies every group.
+	// +optional
+	APIGroups []string
+	// ResourceKinds is a list of kinds of resources within the listed APIGroups this rule applies to.
+	// A ResourceKind can be matched in any of the APIGroups listed above.
+	// Any empty list implies every resource kind.
+	// +optional
+	ResourceKinds []string
```

Comment:
Will do, but I'm going to go with the suggestion of having multiple kinds per group, e.g.:
```go
type GroupKinds struct {
    Group string
    Kinds []string
```

QUESTION: Should `ResourceNames` be moved into this object? I feel like `ResourceNames` faces the same problems as kinds to groups. The "correct" approach seems like it would be to replace `Kinds` with a `Resource struct{ Kind string, ResourceNames []string }`, but I'm concerned that's too heavy weight for now. Can we just remove `ResourceNames` for now?